gem/impero_utils.py

```diff
@@ -241,7 +241,7 @@ def recurse_block(expr, loop_indices):
             refcount.update(recurse(statement, loop_indices))
 
         # Visit :class:`collections.Counter` in deterministic order
-        for e in sorted(refcount.keys(), key=temporaries.index):
+        for e in refcount.keys():
```

Comment:
So here you used to sort the keys which temporaries.index, but that comparison key is `O(n)` since index must traverse the list of temporaries.